### ISSUE TITLE
Improve the way we determine the path for Excel and add-in platform, for debugging

### ISSUE THREAD

**Issue by caioproiete** [Excel-DNA/ExcelDna#19]
This is somewhat related to issue #18, where I'm proposing that we move the Excel.exe path and add-in to the project file, so that the debugging information is persisted in source control, and any developer that clones the repository would then be able to debug the add-in without having to configure anything.

We know that we can't control what version of Excel each developer has installed on his/her machine... Some developers run Excel 2010 32-bit, others run Excel 2010 64-bit, others Excel 2013 64-bit, etc.

I'd like a developer to be able to clone my repository, open my solution in Visual Studio, and just hit F5 and be able to Debug it, starting an instance of the Excel installed on his/her machine, no matter what version is there, and with the right add-in XLL file (x64 or x86) for that debugging session.

---

Whilst I don't have a perfect solution for that yet, I do have a workaround that have been working well for most cases:
Inside the project file, I add a few conditional property groups that define the `StartProgram` variable with the correct path of the Excel on the machine (if matches any, of course)

```
<PropertyGroup Condition="'$(StartProgram)' == '' AND Exists('C:\Program Files\Microsoft Office\Office15\EXCEL.EXE')">
  <StartProgram>C:\Program Files\Microsoft Office\Office15\EXCEL.EXE</StartProgram>
</PropertyGroup>
<PropertyGroup Condition="'$(StartProgram)' == '' AND Exists('C:\Program Files\Microsoft Office\Office14\EXCEL.EXE')">
  <StartProgram>C:\Program Files\Microsoft Office\Office14\EXCEL.EXE</StartProgram>
</PropertyGroup>
<!-- etc... -->
```

I also have the similar conditionals for 32-bit Excel installed on Windows 64-bit machines:

```
<PropertyGroup Condition="'$(StartProgram)' == '' AND Exists('C:\Program Files (x86)\Microsoft Office\Office15\EXCEL.EXE')">
  <StartProgram>C:\Program Files (x86)\Microsoft Office\Office15\EXCEL.EXE</StartProgram>
</PropertyGroup>
<PropertyGroup Condition="'$(StartProgram)' == '' AND Exists('C:\Program Files (x86)\Microsoft Office\Office14\EXCEL.EXE')">
  <StartProgram>C:\Program Files (x86)\Microsoft Office\Office14\EXCEL.EXE</StartProgram>
</PropertyGroup>
<!-- etc... -->
```

And a similar kind of test is applied to determine the right add-in to execute (32-bit or 64-bit)

```
  <PropertyGroup Condition="'$(StartProgram)' != '' AND $(StartProgram.Contains('Program Files (x86)'))">
    <StartArguments>"MyXL-AddIn.xll"</StartArguments>
  </PropertyGroup>
  <PropertyGroup Condition="'$(StartProgram)' != '' AND (! $(StartProgram.Contains('Program Files (x86)')) )">
    <StartArguments>"MyXL-AddIn64.xll"</StartArguments>
  </PropertyGroup>
```

And finally, conditional to set the `StartAction` property

```
<PropertyGroup Condition="'$(StartProgram)' != ''">
  <StartAction>Program</StartAction>
</PropertyGroup>
```

---

**What are your thoughts on this** (or something like this)?

ps: These MSBuild instructions don't necessarily need to be on the `.csproj` (could be a custom .targets file added to the project to make it easier to edit from within Visual Studio).

**Comment by govert** [comment]
Would it be possible to read the registry from the build file to set that property?

If so we can emulate what the Install.ps1 script does to get the right path:
    `$exeValue = Get-ItemProperty -Path Registry::HKEY_CLASSES_ROOT\Excel.XLL\shell\Open\command -name "(default)"`

**Comment by CaioProiete** [comment]
@govert: That's a good idea. I'll investigate if we can read registry entries via MSBuild conditions or properties.

**Comment by govert** [comment]
That would be great. You certainly can't assume that an Excel installation in `C:\Program Files\Microsoft Office\..." is 64-bit. If Windows itself is 32-bit, then the 32-bit Excel will be in this same location.

**Comment by govert** [comment]
Can we merge this with #18 into a single issue, or should they be kept separate?

**Comment by caioproiete** [comment]
Maybe. They can actually be resolved separately, but if I'm going to be the one doing it, I'm probably bundling this and #18 together.

I'll leave it to you. I've been opening these issues primarily to see if I get your buy in, but also secretly hoping that others would want to take a stab on it (or even yourself) and resolve it before I do.

Anyway, both of them will require changing the .targets file, so until #17 is closed, there's not much anyone can do in either.

**Comment by govert** [comment]
OK - I'll leave it open for now.

**Comment by garethhayter** [comment]
I have code to determine the bitness of the running instance of Excel, if you'd find it useful.

**Comment by caioproiete** [comment]
@garethhayter: Thanks. This PR is about launching Excel (independently of having other instances already running), so we first need to determine what is the Excel version installed (and where (path)) and after that, determining the bitness.

If you have any ideas, they are welcome!

**Comment by garethhayter** [comment]
Run the FormulaDesk installer. It displays all the currently installed versions of Excel together with their bitness. Internally it gets the Excel.exe paths too.
Is that what you're looking for? I can give you the code to do that. 


Sent from my Samsung Galaxy smartphone.-------- Original message --------From: Caio Proiete <notifications@github.com> Date: 8/04/2016  09:57  (GMT+12:00) To: gareth.hayter@slyce.com Subject: Re: [Excel-DNA/ExcelDna] Improve the way we determine the path for
  Excel and add-in platform, for debugging (#19) 
@garethhayter: Thanks. This PR is about launching Excel (independently of having other instances already running), so we first need to determine what is the Excel version installed (and where (path)) and after that, determining the bitness.

If you have any ideas, they are welcome!

**Comment by caioproiete** [comment]
@garethhayter: Yes please. If you can past the link to a Gist that would be great.

**Comment by garethhayter** [comment]
@caioproiete: Here is the gist:[https://gist.github.com/garethhayter/66bebda683ac14f3a191fc563eb8aef0](https://gist.github.com/garethhayter/66bebda683ac14f3a191fc563eb8aef0)

**Comment by caioproiete** [comment]
Thanks @garethhayter! This is very useful. As soon as #17 gets merged, I'll see if I can incorporate your code as part of the build.

**Comment by caioproiete** [comment]
I've started to look at this one again, and here are some thoughts:

In order to improve the way we determine the path for Excel and add-in platform, for debugging, we need to:

**1.** Get the installation path of the most recent version of Excel installed on the machine (if any)
**2.** Determine if the most recent version of Excel installed is 32-bit o 64-bit, so we can decide which flavor of the add-in to use when running/debugging
**3.** Allow user to force/override path of a different Excel to use

Does that make sense?

---

On my tests and after looking at @garethhayter's code (thanks again!), I noticed that:
- Excel **64-bit** always writes to the **64-bit hive**
- Excel **32-bit** always writes to **32-bit hive** (even when running on Windows **64-bit**)

With that said, it seems it we could do something very simple like this (_pseudo-code_):

```
var excelInstallKey = Registry("HKLM\Software\Microsoft\Office\?\Excel\InstallRoot", *Hive64*);
if (excelInstallKey.GetValue("Path") != null)
{
    // Found Excel, and it is definitely Excel 64-bit
    return;
}

excelInstallKey = Registry("HKLM\Software\Microsoft\Office\?\Excel\InstallRoot", *Hive32*);
if (excelInstallKey.GetValue("Path") != null)
{
    // Found Excel, and it is definitely Excel 32-bit
    return;
}

// Did not find Excel installed on the machine
```

---

Based on [this StackOverflow answer](http://stackoverflow.com/a/244580/211672), checking the `Path` property on `HKLM\Software\Microsoft\Office\?\Excel\InstallRoot`, where `?` is the number of the Excel version, will work for all Excel versions since Excel 98.

---

### Any thoughts?

---

ps: I see @garethhayter is going the extra mile and checking the PE header of EXCEL.EXE to determine bitness but, based on my tests, checking which registry hive has the Excel path is enough evidence to determine Excel's bitness

**Comment by caioproiete** [comment]
> ps: I see @garethhayter is going the extra mile and checking the PE header of EXCEL.EXE to determine bitness but, based on my tests, checking which registry hive has the Excel path is enough evidence to determine Excel's bitness

After tests, I realized my theory above doesn't work... Unfortunately, opening the 64-bit registry view from a 32-bit machine returns the 32-bit view instead of `null` - which is what I would expect.

Therefore, we'll need to check the PE header of EXCEL.EXE to reliably determine bitness.

**Comment by caioproiete** [comment]
This is resolved by PR #147
